File: apps/caramba-client/packages/caramba_vpn/lib/src/csm/armor.dart
import 'dart:typed_data';

import 'package:caramba_vpn/src/csm/codec/base32_crockford.dart';
import 'package:caramba_vpn/src/csm/errors.dart';
import 'package:caramba_vpn/src/csm/frame.dart';
import 'package:caramba_vpn/src/csm/ids.dart';

// Бронированная текстовая форма и разбиение под QR, 03-WIRE.md 10.
//
//   CARCAP1.<bid>.<i>/<n>.<data>
//
// Любой отказ читателя это E_PARSE_FRAMING. Семейства E_ARMOR_* нет и заводить
// его нельзя: вызывающему все шесть условий означают одно и то же, перед ним не
// поток кадров, а различать их значило бы дать трём реализациям ещё шесть
// поводов разойтись без всякой операционной пользы.

const int csmArmorChunkBytes = 620;
const int csmArmorStreamMax = 65536;
const int csmArmorMaxChunks = 106;
const int csmArmorMaxFrames = 16;

class _Line {
  const _Line(this.bid, this.index, this.count, this.data);
  final String bid;
  final int index;
  final int count;
  final String data;
}

Never _armorFail(String detail) =>
    csmFail(CsmErrorCode.parseFraming, 'armor', detail);

_Line _parseLine(String raw) {
  final line = raw.replaceAll(RegExp(r'\s'), '');
  if (line.length < 21) {
    _armorFail('armored line is too short');
  }
  if (line.substring(0, 7).toUpperCase() != 'CARCAP1' || line[7] != '.') {
    _armorFail('armored line does not begin with the CARCAP1 prefix');
  }
  final afterBid = line.indexOf('.', 8);
  if (afterBid != 16) {
    _armorFail('bid field is not exactly 8 characters');
  }
  final bid = line.substring(8, 16);
  final slash = line.indexOf('/', 17);
  if (slash < 0) {
    _armorFail('armored line carries no ordinal separator');
  }
  final afterCount = line.indexOf('.', slash + 1);
  if (afterCount < 0) {
    _armorFail('armored line carries no data separator');
  }
  final indexText = line.substring(17, slash);
  final countText = line.substring(slash + 1, afterCount);
  final index = _decimal(indexText);
  final count = _decimal(countText);
  if (index < 1 || count < 1 || count > csmArmorMaxChunks || index > count) {
    _armorFail('ordinal $index of $count is outside the legal range');
  }
  return _Line(bid, index, count, line.substring(afterCount + 1));
}

int _decimal(String text) {
  if (text.isEmpty || text.length > 3) {
    _armorFail('ordinal field is empty or too long');
  }
  if (text.length > 1 && text[0] == '0') {
    _armorFail('ordinal field carries a leading zero');
  }
  for (final c in text.codeUnits) {
    if (c < 0x30 || c > 0x39) {
      _armorFail('ordinal field is not decimal');
    }
  }
  return int.parse(text);
}

/// Читает набор бронированных строк и возвращает поток кадров. Порядок строк
/// произвольный; пропущенный порядковый номер обнаруживается, а не
/// склеивается молча.
Uint8List csmArmorDecode(Iterable<String> lines) {
  final parsed = <_Line>[];
  for (final raw in lines) {
    if (raw.trim().isEmpty) {
      continue;
    }
    parsed.add(_parseLine(raw));
  }
  if (parsed.isEmpty) {
    _armorFail('no armored lines');
  }

  final bid = parsed.first.bid;
  final count = parsed.first.count;
  final slices = <int, Uint8List>{};
  for (final line in parsed) {
    if (line.bid != bid || line.count != count) {
      _armorFail('armored set disagrees on bid or on the chunk count');
    }
    final raw = base32CrockfordDecode(line.data);
    if (raw == null) {
      _armorFail(
        'chunk ${line.index} carries a non-Crockford character or non-zero '
        'trailing pad bits',
      );
    }
    if (line.index != count && raw.length != csmArmorChunkBytes) {
      _armorFail(
        'non-final chunk ${line.index} decodes to ${raw.length} bytes where '
        '$csmArmorChunkBytes is required',
      );
    }
    final existing = slices[line.index];
    if (existing != null && !csmBytesEqual(existing, raw)) {
      _armorFail('ordinal ${line.index} appears twice with different data');
    }
    slices[line.index] = raw;
  }
  for (var i = 1; i <= count; i++) {
    if (!slices.containsKey(i)) {
      _armorFail('ordinal $i of $count is missing');
    }
  }

  final joined = <int>[];
  for (var i = 1; i <= count; i++) {
    joined.addAll(slices[i]!);
  }
  if (joined.length > csmArmorStreamMax) {
    _armorFail('frame stream exceeds the $csmArmorStreamMax byte cap');
  }
  final stream = Uint8List.fromList(joined);
  // Сверка bid делается ДО разбора: без неё сканер не отличит смешанные чанки
  // двух разных наборов, а это самый реальный отказ на практике.
  if (csmBundleId(stream) != bid) {
    _armorFail('bid does not match base32_crockford(sha256(stream)[0..5])');
  }
  return stream;
}

/// Читает бронированный текст целиком и возвращает уже разобранные кадры.
List<Uint8List> csmArmorDecodeFrames(String text) {
  final stream = csmArmorDecode(text.split('\n'));
  final frames = csmSplitFrames(stream);
  if (frames.length > csmArmorMaxFrames) {
    _armorFail('frame stream carries ${frames.length} frames, cap is '
        '$csmArmorMaxFrames');
  }
  return frames;
}

/// Кодирование потока кадров в бронированные строки.
List<String> csmArmorEncode(Uint8List stream) {
  if (stream.length > csmArmorStreamMax) {
    _armorFail('frame stream exceeds the $csmArmorStreamMax byte cap');
  }
  final bid = csmBundleId(stream);
  var count = (stream.length + csmArmorChunkBytes - 1) ~/ csmArmorChunkBytes;
  if (count == 0) {
    count = 1;
  }
  if (count > csmArmorMaxChunks) {
    _armorFail('frame stream needs $count chunks, cap is $csmArmorMaxChunks');
  }
  final out = <String>[];
  for (var i = 1; i <= count; i++) {
    final low = (i - 1) * csmArmorChunkBytes;
    var high = low + csmArmorChunkBytes;
    if (high > stream.length) {
      high = stream.length;
    }
    final data = base32CrockfordEncode(stream.sublist(low, high));
    out.add('CARCAP1.$bid.$i/$count.$data');
  }
  return out;
}

File: libs/caramba-core/csm/armor.go
package csm

import (
	"bytes"
	"fmt"
	"strings"
)

// Армированная текстовая форма и разбивка под QR, 03-WIRE.md раздел 10.
//
//	CARCAP1.<bid>.<i>/<n>.<data>
//
// Каждый отказ армированного читателя отображается в E_PARSE_FRAMING. Семейства
// E_ARMOR_* не существует и заводить его нельзя: для вызывающего все шесть
// условий раздела 10.3 означают одно и то же, перед ним не поток кадров, а
// конкретное условие место в журнале, а не в коде.

const (
	// ArmorChunkBytes это 620 байт, то есть 4960 бит, ровно 992 символа
	// base32 без битов заполнения.
	ArmorChunkBytes = 620
	// ArmorMaxChunks следует из потолка потока в 65536 байт.
	ArmorMaxChunks = 106
	armorPrefix    = "CARCAP1"
	// ArmorMediaType это тип содержимого армированного артефакта.
	ArmorMediaType = "text/vnd.caramba.csm1-armor"
)

func armorErr(format string, args ...any) *Error {
	return errf(EParseFraming, "10.3", format, args...)
}

// ArmorEncode кодирует поток кадров в набор строк CARCAP1.
func ArmorEncode(stream []byte) ([]string, error) {
	if len(stream) == 0 {
		return nil, armorErr("empty frame stream")
	}
	if len(stream) > MaxStreamBytes {
		return nil, armorErr("frame stream of %d bytes exceeds the %d byte cap", len(stream), MaxStreamBytes)
	}
	bid := BundleID(stream)
	n := (len(stream) + ArmorChunkBytes - 1) / ArmorChunkBytes
	if n > ArmorMaxChunks {
		return nil, armorErr("%d chunks exceed the %d chunk cap", n, ArmorMaxChunks)
	}
	lines := make([]string, 0, n)
	for i := 0; i < n; i++ {
		lo := i * ArmorChunkBytes
		hi := lo + ArmorChunkBytes
		if hi > len(stream) {
			hi = len(stream)
		}
		lines = append(lines, fmt.Sprintf("%s.%s.%d/%d.%s",
			armorPrefix, bid, i+1, n, Base32CrockfordEncode(stream[lo:hi])))
	}
	return lines, nil
}

type armorLine struct {
	bid  string
	i    int
	n    int
	data []byte
}

// parseDecimal читает десятичное число без ведущих нулей.
func parseDecimal(s string) (int, bool) {
	if s == "" || len(s) > 3 {
		return 0, false
	}
	if len(s) > 1 && s[0] == '0' {
		return 0, false
	}
	v := 0
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return 0, false
		}
		v = v*10 + int(s[i]-'0')
	}
	return v, true
}

func parseArmorLine(line string) (armorLine, error) {
	var out armorLine
	line = strings.TrimSpace(line)
	parts := strings.SplitN(line, ".", 4)
	if len(parts) != 4 {
		return out, armorErr("line does not have the four CARCAP1 fields")
	}
	if !strings.EqualFold(parts[0], armorPrefix) {
		return out, armorErr("line does not start with %s", armorPrefix)
	}
	if len(parts[1]) != 8 {
		return out, armorErr("bid is %d characters, must be 8", len(parts[1]))
	}
	out.bid = strings.ToUpper(parts[1])

	ord := strings.SplitN(parts[2], "/", 2)
	if len(ord) != 2 {
		return out, armorErr("ordinal field is not <i>/<n>")
	}
	i, ok := parseDecimal(ord[0])
	if !ok {
		return out, armorErr("chunk ordinal %q is not a decimal without leading zeros", ord[0])
	}
	n, ok := parseDecimal(ord[1])
	if !ok {
		return out, armorErr("chunk count %q is not a decimal without leading zeros", ord[1])
	}
	if i < 1 || n < 1 || n > ArmorMaxChunks || i > n {
		return out, armorErr("ordinal %d/%d is outside 1..%d", i, n, ArmorMaxChunks)
	}
	out.i, out.n = i, n

	// Пробелы, включая переносы строк, снимаются перед декодированием;
	// дефисы игнорируются декодером Crockford.
	data := strings.Map(func(r rune) rune {
		if r == ' ' || r == '\t' || r == '\r' || r == '\n' {
			return -1
		}
		return r
	}, parts[3])
	b, err := Base32CrockfordDecode(data)
	if err != nil {
		return out, armorErr("chunk %d: %v", i, err)
	}
	out.data = b
	return out, nil
}

// ArmorDecode собирает набор строк CARCAP1 обратно в поток кадров.
// Строки принимаются в любом порядке.
func ArmorDecode(lines []string) ([]byte, error) {
	parsed := make(map[int]armorLine)
	var bid string
	var total int

	seen := 0
	for _, raw := range lines {
		if strings.TrimSpace(raw) == "" {
			continue
		}
		l, err := parseArmorLine(raw)
		if err != nil {
			return nil, err
		}
		if seen == 0 {
			bid, total = l.bid, l.n
		} else {
			if l.bid != bid {
				return nil, armorErr("chunk %d carries bid %s, the set carries %s", l.i, l.bid, bid)
			}
			if l.n != total {
				return nil, armorErr("chunk %d claims a total of %d, the set claims %d", l.i, l.n, total)
			}
		}
		seen++
		if prev, ok := parsed[l.i]; ok {
			// Повторное сканирование идентичной строки допустимо.
			if !bytes.Equal(prev.data, l.data) {
				return nil, armorErr("ordinal %d appears twice with different data", l.i)
			}
			continue
		}
		parsed[l.i] = l
	}
	if seen == 0 {
		return nil, armorErr("no CARCAP1 lines")
	}

	var stream []byte
	for i := 1; i <= total; i++ {
		l, ok := parsed[i]
		if !ok {
			return nil, armorErr("ordinal %d of %d is missing", i, total)
		}
		if i != total && len(l.data) != ArmorChunkBytes {
			return nil, armorErr("non-final chunk %d decodes to %d bytes, must be exactly %d",
				i, len(l.data), ArmorChunkBytes)
		}
		if i == total && (len(l.data) == 0 || len(l.data) > ArmorChunkBytes) {
			return nil, armorErr("final chunk decodes to %d bytes", len(l.data))
		}
		stream = append(stream, l.data...)
	}

	// Проверка bid идёт ДО разбора: набор, смешавший две связки, отвергается
	// раньше, чем его байты попадут в парсер кадров.
	if got := BundleID(stream); got != bid {
		return nil, armorErr("recomputed bid %s does not match the declared %s", got, bid)
	}
	return stream, nil
}

// ArmorDecodeText принимает содержимое файла .carcap целиком.
func ArmorDecodeText(text string) ([]byte, error) {
	return ArmorDecode(strings.Split(text, "\n"))
}
